templates/modern/src/search.ts

```diff
@@ -41,7 +47,8 @@ export function enableSearch() {
 
   function onSearchQueryInput() {
     query = searchQuery.value
-    if (query.length < 3) {
+
+    if (!query || (query.length < 3 && textEncoder.encode(query).length < 3)) {
```

Comment:
```csharp
    if (!query || (query.length < 3 && textEncoder.encode(query).length < 3)) {
```

Could you help me understand the purpose of this change?

Reply:
> if (!query || (query.length < 3 && textEncoder.encode(query).length < 3)) {

### `!query`
It seems not required in current implementation.  I'll remove this code.  
(This check is required when getting input text from `event.data` that passed as argument)

### `query.length < 3`
It is required to keep original behavior and to skip TextEncoder API call if it's not necessary case(query.length>=3). 

### `textEncoder.encode(query).length < 3`
Encode text to UTF-8 bytes and get byte length.
UTF-8 represents single char with 1-3 bytes (Not exactly, but almost cases) 

`< 3` search start condition is determined to support `single character search` for 3-byte chars (e.g. [CJK character](https://en.wikipedia.org/wiki/CJK_characters)).
For these language. It's meaningful to search with single or double character.

2-byte characters(e.g. Latin/Arabic alphabet) seems not be useful to enable single character search.
so I've explicitly set condition to `< 3`.
As a side effect. When 2-byte characters is used. search starts when 2 chars are inputted thought.

Reply:
I've removed unused `!query` condition.

And summarize search start condition.

| Char category       |  Search start condition         |
|---------------------|--------------------------------|
| ASCII                      | At least 3-chars                    |
| UTF-8 2byte char  | At least 2-chars                    |
| UTF-8 3byte char  | single char                            |

Anyway current `lunr.js` based search don't support  non-ASCII text search.
This PR changes don't affect docfx's search behavior.

Reply:
Do you think it make sense to start searching immediately when there is a char, regardless of language?

Reply:
Is it means changing start search condition to `if (query.length == 0)` ?

If it's acceptable to change existing behavior. (At least 3-chars)
It's more simpler solution.
I'll commit above changes and re-test.

There is concern about performance when too much results hit with single char.
As far as I've tested with Azure docs content (About 2000 HTML files generated)
https://github.com/MicrosoftDocs/azure-docs

It's seems acceptable performance
.
In that case. **index build time** and `search-worker.min.js`'s memory usage is problematic.
 `lunr.js` based search are enabled about 5 seconds after page load, and search-worker consume memory about 225MB.

It might be better to introduce prebuild index discussed at https://github.com/dotnet/docfx/issues/661

Reply:
> In that case. index build time and search-worker.min.js's memory usage is problematic.

Is this a regression introduced by changing `query.length < 3` to `query.length == 0` or a known issue discussed at #661 ? Assuming the latter, we could change the "at least 3 chars" behavior in this PR.

Reply:
It's separated issue.and not related in this PR's changes.
I'll commit above changes later.

Reply:
Change search start condition to `if (query.length == 0)` .  (It is equivalent to `if(query === '')`)

And when testing search with large docset.
Sometimes It failed to hide search results. when query is empty.

It is caused by event handling orders. when search operation take times..
Empty string's event is processed first. then search result is rendered.
So I've added additional check to hide search results when query is empty.

Reply:
Note: this PR modify `modern` template search functionality only.
`default` template's search(`default\styles\docfx.js`) script is not changed.